coroutines-extra/src/main/scala/org/coroutines/extra/AsyncAwait.scala

```diff
@@ -0,0 +1,93 @@
+package org.coroutines.extra
+
+
+
+import org.coroutines._
+import scala.annotation.unchecked.uncheckedVariance
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent._
+import scala.language.experimental.macros
+import scala.reflect.macros.whitebox.Context
+import scala.util.{ Success, Failure }
+
+
+
+object AsyncAwait {
+  /** Await the result of a future.
+   *
+   *  When called inside an `async` body, this function will block until its
+   *  associated future completes.
+   *
+   *  @return A coroutine that yields a tuple. `async` will assign this tuple's
+   *          second element to hold the completed result of the `Future` passed
+   *          into the coroutine. The coroutine will directly return the
+   *          result of the future.
+   */
+  def await[R]: Future[R] ~~> (Future[R], R) =
+    coroutine { (awaitedFuture: Future[R]) =>
+      yieldval(awaitedFuture)
+      var result: R = null.asInstanceOf[R]
+      awaitedFuture.value match {
+        case Some(Success(x)) => result = x
+        case Some(Failure(error)) => throw error
+        case None => sys.error("Future was not completed")
+      }
+      result
+    }
+
+  /** Calls `body`, blocking on any calls to `await`.
+   *
+   *  @param  body A coroutine to be invoked.
+   *  @return A `Future` wrapping the result of the coroutine. The future fails
+   *          if `body.hasException` or if one of `body`'s yielded tuples
+   *          has a failing future.
+   */
+  def asyncCall[Y, R](body: ~~~>[Future[Y], R]): Future[R] = {
+    val c = call(body())
+    val p = Promise[R]
+    def loop() {
+      if (!c.resume) {
+        c.tryResult match {
+          case Success(result) => p.success(result)
+          case Failure(exception) => p.failure(exception)
+        }
+      } else {
+        val awaitedFuture = c.value
+        awaitedFuture onComplete {
+          case Success(result) =>
+            loop()
+          case Failure(exception) =>
+            p.failure(exception)
+        }
+      }
+    }
+    Future { loop() }
+    p.future
+  }
+
+  /** Wraps `body` inside a coroutine and asynchronously invokes it using `asyncMacro`.
+   *
+   *  @param  body The block of code to wrap inside an asynchronous coroutine.
+   *  @return A `Future` wrapping the result of `body`.
+   */
+  def async[Y, R](body: =>R): Future[R] = macro asyncMacro[Y, R]
+
+  /** Implements `async`.
+   *
+   *  Wraps `body` inside a coroutine and calls `asyncCall`.
+   *
+   *  @param  body The function to be wrapped in a coroutine.
```

Comment:
How are you doing with Python? Would you consider extending this tool:

https://github.com/storm-enroute/dev-tools/blob/master/lint

with a case that checks that doc comment sections are properly indented?